File: app/src/main/java/igrek/songbook/service/layout/search/SongSearchLayoutController.java
package igrek.songbook.service.layout.search;

import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ImageButton;

import java.util.ArrayList;

import javax.inject.Inject;

import dagger.Lazy;
import igrek.songbook.R;
import igrek.songbook.dagger.DaggerIoc;
import igrek.songbook.logger.Logger;
import igrek.songbook.logger.LoggerFactory;
import igrek.songbook.service.activity.ActivityController;
import igrek.songbook.service.layout.LayoutController;
import igrek.songbook.service.navmenu.NavigationMenuController;
import igrek.songbook.service.songtree.SongTreeItem;
import igrek.songbook.view.songselection.SongListView;

public class SongSearchLayoutController {
	
	@Inject
	Lazy<ActivityController> activityController;
	@Inject
	LayoutController layoutController;
	@Inject
	AppCompatActivity activity;
	@Inject
	NavigationMenuController navigationMenuController;
	
	private Logger logger = LoggerFactory.getLogger();
	private ActionBar actionBar;
	private SongListView itemsListView;
	private Integer scrollPos;
	
	public SongSearchLayoutController() {
		DaggerIoc.getFactoryComponent().inject(this);
	}
	
	public void showSongSearch(View layout) {
		// toolbar
		Toolbar toolbar1 = layout.findViewById(R.id.toolbar1);
		activity.setSupportActionBar(toolbar1);
		actionBar = activity.getSupportActionBar();
		if (actionBar != null) {
			actionBar.setDisplayHomeAsUpEnabled(false);
			actionBar.setDisplayShowHomeEnabled(false);
		}
		
		ImageButton navMenuButton = layout.findViewById(R.id.navMenuButton);
		navMenuButton.setOnClickListener(v -> navigationMenuController.navDrawerShow());
		
		itemsListView = layout.findViewById(R.id.filesList);
		itemsListView.init(activity);
		
		itemsListView.setItems(new ArrayList<>());
	}
	
	public void scrollToItem(int position) {
		itemsListView.scrollTo(position);
	}
	
	public Integer getCurrentScrollPos() {
		return itemsListView.getCurrentScrollPosition();
	}
	
	public void scrollToPosition(int y) {
		itemsListView.scrollToPosition(y);
	}
	
	public void restoreScrollPosition() {
		if (scrollPos != null) {
			scrollToPosition(scrollPos);
		}
	}
	
	public void onItemClickedEvent(int posistion, SongTreeItem item) {
		scrollPos = getCurrentScrollPos();
		logger.debug("item clicked: " + item.getSimpleName());
	}
}

File: app/src/main/java/igrek/songbook/service/layout/search/SongSearchLayoutController.kt
package igrek.songbook.service.layout.search

import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.TextView
import igrek.songbook.R
import igrek.songbook.dagger.DaggerIoc
import igrek.songbook.domain.songsdb.SongsDb
import igrek.songbook.service.layout.LayoutState
import igrek.songbook.service.layout.MainLayout
import igrek.songbook.service.layout.SongSelectionLayoutController
import igrek.songbook.service.songtree.SongTreeFilter
import igrek.songbook.service.songtree.SongTreeItem
import igrek.songbook.service.system.SoftKeyboardService
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.subjects.PublishSubject
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class SongSearchLayoutController : SongSelectionLayoutController(), MainLayout {

    private var searchFilterEdit: EditText? = null
    private var searchFilterSubject: PublishSubject<String> = PublishSubject.create()
    private var itemNameFilter: String? = null
    private var storedScroll: Int? = null

    @Inject
    lateinit var softKeyboardService: SoftKeyboardService

    init {
        DaggerIoc.getFactoryComponent().inject(this)
    }

    override fun showLayout(layout: View) {
        initSongSelectionLayout(layout)

        searchFilterEdit = layout.findViewById(R.id.searchFilterEdit)
        searchFilterEdit!!.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {}
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                searchFilterSubject.onNext(s.toString())
            }
        })
        // refresh only after some inactive time
        searchFilterSubject.debounce(500, TimeUnit.MILLISECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { _ -> setSongFilter(searchFilterEdit!!.text.toString()) }
        if (isFilterSet()) {
            searchFilterEdit!!.setText(itemNameFilter, TextView.BufferType.EDITABLE)
        }
        searchFilterEdit!!.requestFocus()

        itemsListView!!.init(activity, this)
        updateSongItemsList()

        songsDbRepository.dbChangeSubject.subscribe { _ ->
            if (layoutController.isState(LayoutState.SONGS_TREE))
                updateSongItemsList()
        }
    }

    override fun getLayoutState(): LayoutState {
        return LayoutState.SEARCH_SONG
    }

    override fun getLayoutResourceId(): Int {
        return R.layout.song_search
    }

    override fun updateSongItemsList() {
        super.updateSongItemsList()
        // restore Scroll Position
        if (storedScroll != null) {
            itemsListView?.scrollToPosition(storedScroll!!)
        }
    }

    private fun setSongFilter(itemNameFilter: String?) {
        this.itemNameFilter = itemNameFilter
        if (itemNameFilter == null)
            searchFilterEdit?.setText("", TextView.BufferType.EDITABLE)
        // reset scroll
        storedScroll = null
        updateSongItemsList()
    }

    override fun getSongItems(songsDb: SongsDb): List<SongTreeItem> {
        // no filter
        if (!isFilterSet()) {
            return songsDb.getAllUnlockedSongs()
                    .map { song -> SongTreeItem.song(song) }
        } else {
            val songNameFilter = SongTreeFilter(itemNameFilter)
            return songsDb.getAllUnlockedSongs()
                    .map { song -> SongTreeItem.song(song) }
                    .filter { item -> songNameFilter.matchesNameFilter(item) }
        }
    }

    private fun isFilterSet(): Boolean {
        return itemNameFilter != null && itemNameFilter!!.isEmpty()
    }

    fun onBackClicked() {
        if (isFilterSet()) {
            setSongFilter(null)
        } else {
            activityController.get().quit()
        }
    }

    override fun onSongItemClick(item: SongTreeItem) {
        // store Scroll Position
        storedScroll = itemsListView!!.currentScrollPosition
        if (item.isSong) {
            openSongPreview(item)
        }
    }

}
